File: python/voxgig_sekreto/plugins/doppler.py
# The doppler plugin: Doppler, one bulk download per config. Needs
# HTTPS. A port of typescript/plugins/doppler.ts.
from ..addr import checkaddr
from ..sekreto import SekretoError, envkey
from ..providers import Provider, providerplugin
from .httpjson import fetchjson, urlpart


class DopplerProvider(Provider):
    """Doppler.

    The whole config is downloaded once - Doppler's own bulk endpoint -
    and answered from memory, like a remote .env: `api.token` is the
    `API_TOKEN` entry. A service token is config-scoped, so project and
    config are only needed with broader tokens.
    """

    def __init__(self, options=None):
        self.opts = options or {}
        self.values = None

    def load(self):
        if self.values is not None:
            return self.values

        addr = (self.opts.get('addr') or 'https://api.doppler.com').rstrip('/')
        checkaddr(addr)

        url = addr + '/v3/configs/config/secrets/download?format=json'
        if self.opts.get('project'):
            url += '&project=' + urlpart(self.opts['project'])
        if self.opts.get('config'):
            url += '&config=' + urlpart(self.opts['config'])

        status, body = fetchjson(
            'GET', url, {'authorization': 'Bearer ' + (self.opts.get('token') or '')}
        )

        if 200 != status or not isinstance(body, dict):
            raise SekretoError('sekreto: doppler error: ' + str(status))

        self.values = {}
        for key, value in body.items():
            if value is not None:
                self.values[key] = str(value)

        return self.values

    def lookup(self, name):
        return self.load().get(envkey(name))

    def describe(self):
        if self.opts.get('project'):
            return 'doppler:' + self.opts['project'] + '/' + (self.opts.get('config') or '')
        return 'doppler'


# The plugin: the `doppler` provider kind, as a voxgig/plugin definition.
doppler = providerplugin('doppler', DopplerProvider)

File: ruby/lib/voxgig_sekreto/plugins/doppler.rb
# frozen_string_literal: true

# The doppler plugin: Doppler. Needs HTTPS. A port of
# typescript/plugins/doppler.ts, which is canonical.

require_relative '../../voxgig_sekreto'
require_relative '../addr'
require_relative 'httpjson'

module VoxgigSekreto
  # Doppler.
  #
  # The whole config is downloaded once - Doppler's own bulk endpoint -
  # and answered from memory, like a remote .env: `api.token` is the
  # `API_TOKEN` entry. A service token is config-scoped, so project and
  # config are only needed with broader tokens.
  class DopplerProvider
    def initialize(opts = nil)
      @opts = opts || {}
      @values = nil
    end

    def load
      return @values unless @values.nil?

      addr = (@opts['addr'] || 'https://api.doppler.com').sub(%r{/\z}, '')
      VoxgigSekreto.checkaddr(addr)

      url = addr + '/v3/configs/config/secrets/download?format=json'
      url += '&project=' + VoxgigSekreto.uriescape(@opts['project']) if VoxgigSekreto.given(@opts['project'])
      url += '&config=' + VoxgigSekreto.uriescape(@opts['config']) if VoxgigSekreto.given(@opts['config'])

      status, body = VoxgigSekreto.fetchjson(
        'GET', url, { 'authorization' => 'Bearer ' + (@opts['token'] || '') }
      )

      raise SekretoError, 'sekreto: doppler error: ' + status.to_s if 200 != status || !body.is_a?(Hash)

      @values = {}
      body.each { |key, value| @values[key] = value.to_s unless value.nil? }

      @values
    end

    def lookup(name)
      load[VoxgigSekreto.envkey(name)]
    end

    def describe
      return 'doppler' unless VoxgigSekreto.given(@opts['project'])

      'doppler:' + @opts['project'] + '/' + (@opts['config'] || '')
    end
  end

  module Plugins
    # The plugin: the `doppler` provider kind, as a voxgig/plugin
    # definition.
    DOPPLER = VoxgigSekreto.providerplugin('doppler', lambda { |spec|
      DopplerProvider.new(spec.slice('token', 'project', 'config', 'addr'))
    })
  end
end
